monitoring/check_collectd_mlab_test.py

```diff
@@ -0,0 +1,379 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for check_collectd_mlab check."""
+
+import contextlib
+import os
+import socket
+import threading
+import time
+import unittest2 as unittest
+
+# Third-party modules.
+import mock
+
+# Module under test.
+import check_collectd_mlab
+
+# R0904: Too many public methods. Hard to avoid for unit tests.
+# pylint: disable=R0904
+
+
+class FakeCollectd(object):
+  """A fake version of collectd that listens to a unix domain socket."""
+
+  def __init__(self, unixsock):
+    self._unixsock = unixsock
+    self._sock = None
+    self._thread = threading.Thread(target=self._serve)
+    self._thread.daemon = True
+    self._failure_mode = ''
+
+  def start(self):
+    """Starts the FakeCollectd service, listening on the unix domain socket."""
+    if os.path.exists(self._unixsock):
+      os.unlink(self._unixsock)
+    self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+    self._sock.bind(self._unixsock)
+    self._sock.listen(1)
+    self._thread.start()
+
+  def shutdown(self):
+    """Waits up to 5 seconds for the FakeCollectd thread to shutdown.
+
+    Returns:
+      bool, True if shutdown succeeds, False if shutdown fails.
+    """
+    self._thread.join(5)
+    return not self._thread.isAlive()
+
+  def set_response_behavior(self, mode):
+    """Sets the mode for injecting errors into the client response.
+
+    If mode is not one of the following, it has no effect.
+      "send_bad_reply" - the status code is not a number.
+      "send_empty_reply" - the request is received but no reply is sent.
+      "close_after_reply" - the connection is closed after sending the
+         number of results but before sending the remaining values.
+
+    Args:
+      mode: str, the fake server failure mode.
+    """
+    self._failure_mode = mode
+
+  def _serve(self):
+    """Runs a fake server that can respond to a single request.
+
+    This fake service emulates GETVAL with the option to inject errors.
+      https://collectd.org/wiki/index.php/Plain_text_protocol#GETVAL
+    """
+    # E1101: attribute access for nonexisting member. pylint 0.21.1 rejects
+    # the calls to sendall. Disable this check because the method is present.
+    # pylint: disable=E1101
+    # Wait for a connection.
+    with contextlib.closing(self._sock):
+      client, _ = self._sock.accept()
+      with contextlib.closing(client):
+        # We only test with the GETVAL command, so request is ignored.
+        _ = check_collectd_mlab.sock_readline(client)
+
+        # Send "number of values" response.
+        if self._failure_mode == 'send_bad_reply':
+          client.sendall('not-a-number junk\n')
+        elif self._failure_mode == 'send_empty_reply':
+          client.sendall('\n')
+        else:
+          client.sendall('1 default reply\n')
+
+        # Send "actual values" response.
+        if self._failure_mode == 'close_after_reply':
```

Comment:
Suggest renaming to `close_during_reply` because it took me a bit looking at 101-108 to figure out why we were saying we were closing _after_ reply when we seemed to be closing _before_ the reply.